File: 1/2021.07.08/103.Binary Tree Zigzag Level Order Traversal/python/solution.py
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        # DFS
        def DFS(root, level):
            if root:
                if len(res) == level:
                    res.append([])
                if level % 2 == 0:
                    res[level].append(root.val)
                else:
                    res[level].insert(0, root.val)
                DFS(root.left, level + 1)
                DFS(root.right, level + 1)
        
        res = []
        DFS(root, 0)
        return res

File: 1/2021.07.08/103.Binary Tree Zigzag Level Order Traversal/java/solution.java
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        // BFS
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Deque<TreeNode> q = new ArrayDeque<>();
        q.offerLast(root);
        boolean flag = true;
        while (!q.isEmpty()) {
            int size = q.size();
            LinkedList<Integer> level = new LinkedList<>();
            while (size-- > 0) {
                TreeNode node = q.pollFirst();
                if (flag) {
                    level.offerLast(node.val);
                } else {
                    level.offerFirst(node.val);
                }
                if (node.left != null) {
                    q.offerLast(node.left);
                }
                if (node.right != null) {
                    q.offerLast(node.right);
                }
            }
            res.add(level);
            flag = !flag;
        }
        return res;
    }
}
